### PR TITLE
[clang] Warn when builtin names are used outside of invocations

### PR DESCRIPTION
**Pull request by cor3ntin**
This patch adds a deprecation warning when builtin names like `__is_pointer` are used for something else than builtin invocation. For instance, up until a week ago libstdc++ declared their own `struct __is_pointer` type. libc++ has been doing the same until 3.5. This will be diagnosed now.

The implementation is as suggested by Richard in https://github.com/llvm/llvm-project/pull/95969#discussion_r1644997251: if lexer sees a revertible type trait builtin identifier that is not followed by an lparen, it issues a warning. Notable corner case of `__has_builtin(__is_pointer)` is exempted from that, because such usage is perfectly fine.

### CODE REVIEW

File: clang/lib/Lex/PPMacroExpansion.cpp [hunk 1]
```diff
@@ -1673,6 +1673,7 @@ void Preprocessor::ExpandBuiltinMacro(Token &Tok) {
         return II && HasExtension(*this, II->getName());
       });
   } else if (II == Ident__has_builtin) {
+    EvaluatingHasBuiltinMacro = true;
```
Comment (Endilll) [inline clang/lib/Lex/PPMacroExpansion.cpp:1676]: I'm not particularly proud of this workaround, and if reviewers think that we should do this in some other way (e.g. enable backtracking for every `__has_builtin` use, and look behind when we see a revertible type trait identifier), I'm open for suggestions.

**Comment by Endilll** [comment]
CC @MitalAshok

**Comment by Endilll** [comment]
To clear potential confusion, Corentin suggested to me offline to take over this PR. I fixed the tests and implemented the workaround for `__has_builtin` case, which surfaced while I was fixing the tests.

**Comment by MitalAshok** [comment]
Looks like `!__is_identifier(<revertible type trait name>)` is used instead of `__has_builtin` sometimes. With this patch, `__is_identifier(<revertible type trait name>)` is true because there's no lparen after the name

In libc++: 7f302f220e7b8727ed1bf8832dcc2d87b897e527 (Pretty easy fix to replace `__has_keyword(__reference_binds_to_temporary)` with `__has_builtin(__reference_binds_to_temporary)` again though)

In WIL which seems to be used by a few people: https://github.com/microsoft/wil/blob/68ab8c19ef557e1006cae6b1b84dedf12c91c9d6/include/wil/wistd_config.h#L249

libstdc++, indirectly through a macro: https://gcc.gnu.org/git/?p=gcc.git;a=blobdiff;f=libstdc%2B%2B-v3/include/bits/c%2B%2Bconfig;h=27302ed392eb163954c5a2d367831dbcb0ead3c3;hp=2e6c880ad95a1cc1dfdead2bbcfb977960172ac5;hb=6aa12274007bccbae2691a9d336c37fe167bb535;hpb=6ea5a23766b8077a503362c4fa6f51de92669c11

These will now all report that those things are identifiers and not available as builtins. It can be supported with the same workaround used for `__has_builtin`. Apparantly `__has_feature(__is_abstract)` is also supposed to be true (https://github.com/llvm/llvm-project/blob/8d4aa1f22ea08b12a7958b681e8a265f78cb349f/clang/docs/LanguageExtensions.rst?plain=1#L1693), but that seems to be broken on main and not a new issue with this patch.

---

Also it doesn't seem ideal to not support things like:

```c++
#define IS_CLASS __is_class
static_assert(!IS_CLASS(void));
```

Since these are supported by GCC and MSVC. Could you check again for an lparen after a macro expansion?

**Comment by Endilll** [comment]
> Looks like !__is_identifier(<revertible type trait name>) is used instead of __has_builtin sometimes. With this patch, __is_identifier(<revertible type trait name>) is true because there's no lparen after the name

Thank you, I'll make `__is_identifier` work.

> Apparantly __has_feature(__is_abstract) is also supposed to be true (...), but that seems to be broken on main and not a new issue with this patch.

I can't make it work even on ancient version of Clang on Compiler Explorer, so I wonder if it ever worked.

> Also it doesn't seem ideal to not support things like:
>
> #define IS_CLASS __is_class
> static_assert(!IS_CLASS(void));
>
> Since these are supported by GCC and MSVC. Could you check again for an lparen after a macro expansion?

I think this would require us to abandon the lex approach implemented here. But read on...

> In libc++: https://github.com/llvm/llvm-project/commit/7f302f220e7b8727ed1bf8832dcc2d87b897e527 (Pretty easy fix to replace __has_keyword(__reference_binds_to_temporary) with __has_builtin(__reference_binds_to_temporary) again though)

Goodness, I think this usage makes lex approach non-viable. Even if libc++ fixes that now, it has been deployed in system headers.

**Comment by Endilll** [comment]
@zygoloid @cor3ntin @AaronBallman in the light of new data in https://github.com/llvm/llvm-project/pull/96097#issuecomment-2198314023, do you think this lex approach to revertible type trait identifiers is still viable?

**Comment by Endilll** [comment]
```
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/uniform_int_dist.h:250:31: error: expected unqualified-id
          static_assert(!_Up_traits::__is_signed, "U must be unsigned");
```
Results from CI are not making this any better.

**Comment by MitalAshok** [comment]
What if we went in the other direction? We want to deprecate `__is_pointer` as an identifier, so only make it an identifier when it is being used by libstdc++ as an identifier. libstdc++ usage looks something like:

```c++
// type template
template<typename T>
struct __is_pointer;
template<typename T>
struct __is_pointer<T*> : true_type {};
__is_pointer<T>::value

// variable template
template<typename T>
inline constexpr bool __is_pointer = ...
__is_pointer<T>
```

So only make it an identifier if it is preceded by `struct`/`class` or if it is succeeded by `<`/`=`?

**Comment by cor3ntin** [comment]
In light of further complexities noted here, I'm abandoning this in favor of #95969

**Pull request closed without merging**